core/src/banking_stage/transaction_scheduler/transaction_packet_container.rs

```diff
@@ -0,0 +1,324 @@
+    pub(crate) transaction: SanitizedTransaction,
+    pub(crate) max_age_slot: Slot,
+}
+
+pub(crate) struct TransactionPacketContainer {
+    /// Unscheduled unique identifiers ordered by priority
+    priority_queue: MinMaxHeap<TransactionPriorityId>,
+    /// Map from `TransactionId` to `SanitizedTransactionTTL`.
+    /// Contains transactions that have not yet been scheduled for consuming.
+    id_to_transaction_ttl: HashMap<TransactionId, SanitizedTransactionTTL>,
+    /// Map from `TransactionId` to `DeserializedPacket`.
+    /// Contains packets for the entire lifetime of packet/transaction.
+    /// These will only be removed once a transaction has finished processing,
+    /// or if the packets are forwarded without holding.
+    id_to_packet: HashMap<TransactionId, DeserializedPacket>,
+}
+
+impl TransactionPacketContainer {
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            priority_queue: MinMaxHeap::with_capacity(capacity),
+            id_to_transaction_ttl: HashMap::with_capacity(capacity),
+            id_to_packet: HashMap::with_capacity(capacity),
+        }
+    }
+
+    /// Returns true if the queue is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.priority_queue.is_empty()
+    }
+
+    /// Returns the remaining capacity of the queue
+    pub(crate) fn remaining_queue_capacity(&self) -> usize {
+        self.priority_queue.capacity() - self.priority_queue.len()
+    }
+
+    /// Get a non-consuming iterator over the top `n` transactions in the queue.
+    pub(crate) fn take_top_n(
+        &mut self,
+        n: usize,
+    ) -> impl Iterator<Item = TransactionPriorityId> + '_ {
+        (0..n).map_while(|_| self.priority_queue.pop_max())
+    }
+
+    /// Get packet by id.
+    pub(crate) fn get_packet(&self, id: &TransactionId) -> Option<&DeserializedPacket> {
+        self.id_to_packet.get(id)
+    }
+
+    /// Get mutable packet by id.
+    pub(crate) fn get_mut_packet(&mut self, id: &TransactionId) -> Option<&mut DeserializedPacket> {
+        self.id_to_packet.get_mut(id)
+    }
+
+    /// Get transaction by id.
+    /// Panics if the transaction does not exist.
+    pub(crate) fn get_transaction(&self, id: &TransactionId) -> &SanitizedTransactionTTL {
+        self.id_to_transaction_ttl
+            .get(id)
+            .expect("transaction must exist")
+    }
+
+    /// Take transaction by id.
+    /// Panics if the transaction does not exist.
+    pub(crate) fn take_transaction(&mut self, id: &TransactionId) -> SanitizedTransactionTTL {
+        self.id_to_transaction_ttl
+            .remove(id)
+            .expect("transaction must exist")
+    }
+
+    /// Insert a new transaction into the container's queues and maps.
+    pub(crate) fn insert_new_transaction(
+        &mut self,
+        transaction_id: TransactionId,
+        packet: ImmutableDeserializedPacket,
+        transaction_ttl: SanitizedTransactionTTL,
```

Comment:
there is a sanitized_transaction in both `packet` and `transaction_ttl`, are they suppose to be the same tx? Can it be `insert_new_transaction(&mut self, packet: ImmutableDeserializedPacket, max_age_slot: Slot)`?

Reply:
Only used here: https://github.com/solana-labs/solana/pull/32981/files#diff-8515945de0b5bae53765b62809c72d3bc2f875b2099efaa03e27345667af2635R214

Would have to move the sanitization and lock validation inside.